src/util/structured_data.cpp

```diff
@@ -0,0 +1,218 @@
+      lower_case.begin(), lower_case.end(), [](char &c) { c = ::tolower(c); });
+    return lower_case;
+  };
+  this->components = make_range(components)
+                       .map(to_lower_string)
+                       .collect<std::vector<std::string>>();
+}
+std::string labelt::camel_case() const
+{
+  std::ostringstream output;
+  if(!components.empty())
+  {
+    output << *components.begin();
+    join_strings(
+      output, std::next(components.begin()), components.end(), "", capitalize);
+  }
+  return output.str();
+}
+std::string labelt::snake_case() const
+{
+  std::ostringstream output;
+  join_strings(output, components.begin(), components.end(), '_');
+  return output.str();
+}
+std::string labelt::kebab_case() const
+{
+  std::ostringstream output;
+  join_strings(output, components.begin(), components.end(), '-');
+  return output.str();
+}
+std::string labelt::pretty() const
+{
+  std::ostringstream output;
+  if(!components.empty())
+  {
+    const auto range =
+      make_range(components.begin(), std::next(components.begin()))
+        .map(capitalize)
+        .concat(make_range(std::next(components.begin()), components.end()));
+
+    join_strings(output, range.begin(), range.end(), ' ');
+  }
+  return output.str();
+}
+bool labelt::operator<(const labelt &other) const
+{
+  return camel_case() < other.camel_case();
+}
+structured_data_entryt structured_data_entryt::data_node(const jsont &data)
+{
+  // Structured data (e.g. arrays and objects) should use an entry
+  PRECONDITION(!(data.is_array() || data.is_object()));
+  return structured_data_entryt(data);
+}
+structured_data_entryt
+structured_data_entryt::entry(std::map<labelt, structured_data_entryt> children)
+{
+  return structured_data_entryt(children);
+}
+structured_data_entryt::structured_data_entryt(const jsont &data) : data(data)
+{
+}
+structured_data_entryt::structured_data_entryt(
+  std::map<labelt, structured_data_entryt> children)
+  : children(std::move(children))
+{
+}
+bool structured_data_entryt::is_leaf() const
+{
+  return children.empty();
+}
+std::string structured_data_entryt::leaf_data() const
+{
+  return data.value;
+}
+std::map<labelt, structured_data_entryt>
+structured_data_entryt::get_children() const
+{
+  return children;
+}
+jsont structured_data_entryt::leaf_object() const
+{
+  return data;
+}
+structured_datat::structured_datat(
+  std::map<labelt, structured_data_entryt> data)
+  : data(std::move(data))
+{
+}
+
+xmlt xml_node(const std::pair<labelt, structured_data_entryt> &entry)
+{
+  const labelt &label = entry.first;
+  const structured_data_entryt &data = entry.second;
+  xmlt output_data{label.kebab_case()};
+  if(data.is_leaf())
+  {
+    output_data.data = data.leaf_data();
+  }
+  else
+  {
+    const auto &children = data.get_children();
+    output_data.elements =
+      make_range(children).map(xml_node).collect<std::list<xmlt>>();
+  }
+  return output_data;
+}
+
+xmlt structured_datat::to_xml() const
```

Comment:
One design question I have (which I honestly don't know what the right answer is :-) ) is - who's responsibility is it to do the conversion from `structured_data_entryt` to {xml,json,plain} ? It feels like the knowledge of what (say) xml needs to look like (e.g. kebab_case) should be more closely located with the various XML utility functions. So perhaps the design should be that structure_data.{h,cpp} has a `to_xm()` `to_json()` and `to_plain()`, and those basically just be simple wrappers around other functions like `xml_node(...)` in the XML utility modules? Then all the knowledge of what XML should look like would be locate in the XML utility modules. Of course that means xml/json/etc modules need to know a little bit about structured_data... so it's not a clear choice either way.